aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,339 @@
+  :lines: 21
+  :dedent: 12
+
+``ServiceFilterAttribute`` implements ``IFilterFactory``, which exposes a single method for creating an ``IFilter`` instance. In the case of ``ServiceFilterAttribute``, the ``IFilterFactory`` interface's ``CreateInstance`` method is implemented to load the specified type from the services container (DI).
+
+``TypeFilterAttribute`` is very similar to ``ServiceFilterAttribute`` (and also implements ``IFilterFactory``), but its type is not resolved directly from the DI container. Instead, it instantiates the type using an `ObjectFactory delegate <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/Extensions/DependencyInjection/ObjectFactory/index.html>`_. Because of this difference, types that are referenced using the ``TypeFilterAttribute`` do not need to be registered with the container first (but they will still have their dependencies fulfilled by the container). Also, ``TypeFilterAttribute`` can optionally accept constructor arguments for the type in question. The following example demonstrates how to pass arguments to a type using ``TypeFilterAttribute``:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/HomeController.cs
+  :language: c#
+  :emphasize-lines: 1-2
+  :lines: 33-39
+  :dedent: 8
+
+If you have a simple filter that doesn't require any arguments, but which has constructor dependencies that need to be filled by DI, you can inherit from ``TypeFilterAttribute``, allowing you to use your own named attribute on classes and methods (instead of ``[TypeFilter(typeof(FilterType))]``). The following filter shows how this can be implemented:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleLoggingActionFilterAttribute.cs
+  :language: c#
+  :emphasize-lines: 1, 3, 7
+  :lines: 8-38
+  :dedent: 4
+
+This filter can be applied to classes or methods using the ``[SampleLoggingActionFilter]`` syntax.
+
+``IFilterFactory`` implements ``IFilter``. Therefore, an ``IFilterFactory`` instance can be used as an ``IFilter`` instance anywhere in the filter pipeline. When the framework prepares to invoke the filter, attempts to cast it to an ``IFilterFactory``. If that cast succeeds, the ``CreateInstance`` method is called to create the ``IFilter`` instance that will be invoked. This provides a very flexible design, since the precise filter pipeline does not need to be set explicitly when the application starts.
+
+You can implement ``IFilterFactory`` on your own attribute implementations as another approach to creating filters:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/AddHeaderWithFactoryAttribute.cs
+  :language: c#
+  :emphasize-lines: 1,4-7
+  :lines: 6-26
+  :dedent: 4
+
+@rynowak - can you clarify this next bit?
+If you're using ``IFilterFactory``, you can specify lifetime of the filter (how?). When your filter is the instance, your filter instance is cached, so avoid do anything stateful in the filter returned by ``CreateInstance``.
```

Comment:
This is an RC2 feature. We should add the details when they are available. In RC1 your filters returned by an `IFilterFactory` are never cached.